src/main/java/org/jenkinsci/remoting/engine/jnlp3/ChannelCiphers.java

```
@@ -0,0 +1,100 @@
+    private final Cipher decryptCipher;
+
+    ChannelCiphers(byte[] aesKey, byte[] specKey, Cipher encryptCipher, Cipher decryptCipher) {
+        this.aesKey = aesKey;
+        this.specKey = specKey;
+        this.encryptCipher = encryptCipher;
+        this.decryptCipher = decryptCipher;
+    }
+
+    public byte[] getAesKey() {
+        return aesKey;
+    }
+
+    public byte[] getSpecKey() {
+        return specKey;
+    }
+
+    public Cipher getEncryptCipher() {
+        return encryptCipher;
+    }
+
+    public Cipher getDecryptCipher() {
+        return decryptCipher;
+    }
+
+    /**
+     * Create a pair of AES symmetric key {@link javax.crypto.Cipher}s using
+     * randomly generated keys.
+     *
+     * @throws IOException If there is a problem constructing the ciphers.
+     */
+    public static ChannelCiphers create() throws IOException {
+        return create(Jnlp3Util.generate128BitKey(), Jnlp3Util.generate128BitKey());
+    }
+
+    /**
+     * Creates a pair of AES symmetric key {@link javax.crypto.Cipher}s using
+     * the given AES key and {@link javax.crypto.spec.IvParameterSpec} key.
+     *
+     * @throws IOException If there is a problem constructing the ciphers.
+     */
+    public static ChannelCiphers create(byte[] aesKey, byte[] specKey) throws IOException {
+        try {
+            SecretKey secretKey = new SecretKeySpec(aesKey, "AES");
+            IvParameterSpec spec = new IvParameterSpec(specKey);
+            Cipher encryptCipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            encryptCipher.init(Cipher.ENCRYPT_MODE, secretKey, spec);
+            Cipher decryptCipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            decryptCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+
+            return new ChannelCiphers(aesKey, specKey, encryptCipher, decryptCipher);
+        } catch (Exception e) {
+            throw new IOException("Failed to create channel ciphers", e);
```

I feel a bit uneasy about ChannelCiphers and HandshakeCiphers throwing IOExceptions. I feel they should throw an Exception and JnlpProtocol.establishChannel should be allowed to bubble up an Exception to Engine, which can be caught and handled.